code/CalDisBetCableLandingPoints.py

```diff
@@ -0,0 +1,315 @@
+    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
+    distance = R * c
+
+    return floatFormatter(distance)
+
+# Helper function to add edge
+
+
+def add_edge(G, point1, point2):
+    # Calculate the distance between the two points
+    distance = haversine(point1, point2)
+    # Add the edge to the graph with distance as weight
+    G.add_edge(point1, point2, weight=distance)
+
+
+def construct_graph_with_networkx(paths):
+    """
+    This function take a list of paths and return the corresponding graph.
+    
+    Parameters:
+    paths(list) : a list of list in the following format [path1, path2 ... pathx]
+    [[node 1, node2 .... nodeX], [node a, node b, node c] .... []].
+    each node has the following format (longtitude, latitude)
+    
+    Returns:
+    G (networkX graph): the graph generated for corresponding path list.
+    
+    
+    """
+    # Create a NetworkX graph
+    G = nx.Graph()
+
+    # Add edges for consecutive points within paths
+    for path in paths:
+        for i in range(len(path) - 1):
+            add_edge(G, path[i], path[i + 1])
+
+    # Connect paths by shared points
+    for path in paths:
+        # all_other_points includes all other points appreared in other paths.
+        all_other_points = set(sum((p for p in paths if p is not path), []))
+        
+        # Only need to consider first and last points for connection
+        for point in [path[0], path[-1]]:
+            point_neighbor_list = [neighbor for neighbor in G.neighbors(
+                point) if neighbor not in path]
+            # if the point also appear in other paths, merge them
+            if point_neighbor_list:
+                for neighbor in point_neighbor_list:
+                    add_edge(G, point, neighbor)
+            # if the point does not appear in other paths, find the closest point to it
+            # if the distance between them smaller than 10 kilometer, then connect them.
+            else:
+                if all_other_points:
+                    # Find the closest point and its distance
+                    closest_point, min_distance = min(
+                        ((other_point, haversine(point, other_point))
+                            for other_point in all_other_points),
+                        key=lambda item: item[1]
+                    )
+                    threshold = 10
+                    # Add edge if the closest distance is under the threshold and not in the current path
+                    if min_distance < threshold and closest_point not in path:
+                        add_edge(G, point, closest_point)
+
+    return G
+
+def verify_graph_with_cities(graph, start_city, end_city, cable_id):
+    """
+    Before calculate the distance between a city pair. This function used to verify whether both of the cities appears in
+    the graph, if not, update the graph if the city is with in the 125 kilometer circle of any submarine cable points. Connect them with a edge with distance weight.
```

Comment:
Instead of updating the graph, I thought we would update the database. Then we can run several setup scripts (adding the submarine cables, adding the cloud regions, etc) only one time. Then we can use the existing shortest path algorithm to find all the paths between pairs without doing all the setup each time.

Reply:
the result already been added to a new table in database called submarine_standard_paths